Birthday-2024-Project/Scripts/GridLogic.gd

```diff
@@ -147,36 +219,65 @@ func SetGridSpacesByPieceShape(gridCoords : Vector2i, newValue : GridSpaceInfo.G
 
 func CheckLegalToPlace(piece : PieceLogic) -> bool:
 	var pieceCoords : Vector2i = PositionToGridCoordinate(piece.GetOriginCellPosition())
-	#if off the grid map
-	if pieceCoords.x < xMinGrid or pieceCoords.x > xMaxGrid:
-		return false
-	if pieceCoords.y < yMinGrid or pieceCoords.y > yMaxGrid:
-		return false
+	
+	#if outside the level bounding box
+	if gridMode == GridMode.PLAY:
+		if pieceCoords.x < xMinGrid or pieceCoords.x > xMaxGrid:
+			return false
+		if pieceCoords.y < yMinGrid or pieceCoords.y > yMaxGrid:
+			return false
+	
+	var cellOffsetsArray : Array[Vector2i] = piece.GetPieceShapeOffsetArray()
+	for shapeCell in cellOffsetsArray:
+		var spaceStatus : GridSpaceInfo = GetGridSpace(pieceCoords + shapeCell)
+		#would go out of max bounds
+		if spaceStatus == null:
+			return false
+		if gridMode == GridMode.PLAY:
+			if spaceStatus.currentStatus != GridSpaceInfo.GridSpaceStatus.OPEN:
+				return false
+		else:
+			if spaceStatus.currentStatus == GridSpaceInfo.GridSpaceStatus.OCCUPIED:
+				return false
+	return true
+
+func CheckIfInDeletionZone(piece : PieceLogic) -> bool:
+	var pieceCoords : Vector2i = PositionToGridCoordinate(piece.GetOriginCellPosition())
 	var cellOffsetsArray : Array[Vector2i] = piece.GetPieceShapeOffsetArray()
 	for shapeCell in cellOffsetsArray:
 		var spaceStatus : GridSpaceInfo = GetGridSpace(pieceCoords + shapeCell)
-		if spaceStatus.currentStatus != GridSpaceInfo.GridSpaceStatus.OPEN:
+		#is out of max bounds
+		if spaceStatus != null:
 			return false
+	#all squares are outside the grid map
 	return true
 
 func PlacePiece(piece : PieceLogic) -> bool:
-	if CheckLegalToPlace(piece) == false:
-		return false
 	var pieceCoords : Vector2i = PositionToGridCoordinate(piece.GetOriginCellPosition())
-	SetGridSpacesByPieceShape(pieceCoords, GridSpaceInfo.GridSpaceStatus.OCCUPIED, piece)
-	piece.AssignMapGridCoordinates(pieceCoords)
-	
-	grid_updated.emit()
+	return PlacePieceByCoordinates(piece, pieceCoords)
+
+func PlacePieceByCoordinates(piece : PieceLogic, gridCoord : Vector2i) -> bool:
+	if gridMode == GridMode.PLAY:
+		if CheckLegalToPlace(piece) == false:
+			return false
+	SetGridSpacesByPieceShape(gridCoord, GridSpaceInfo.GridSpaceStatus.OCCUPIED, piece)
```

Comment:
after leaving edit mode, pieces can cover buttons and be partially outside of the screen (checked on windowed and fullscreen), maybe it would be better to spawn them at the sides of puzzle and not on all sides of it

Reply:
I believe this is also related to issue https://github.com/Saplings-Projects/Birthday-2024/issues/3
since the positions are fixed and don't adjust to other resolutions